### PR TITLE
GH696: Added WiX heat support

### PR DESCRIPTION
**Pull request by RLittlesII**
Implements #696 
Still WIP.  I am going through the doc listed below and trying to hit everything that is not deprecated.  If there are switches that you think are not as relevant please call them out.

http://wixtoolset.org/documentation/manual/v3/overview/heat.html

Switch | Description
------------ | -------------
-ag | Auto generate component guids at compile time, e.g. set Guid="*".
-cg <ComponentGroupName> | Component group name (cannot contain spaces e.g -cg MyComponentGroup).
-configuration | Configuration to set when harvesting the project.
-directoryid | Overridden directory id for generated directory elements.
-dr <DirectoryName> | Directory reference to root directories (cannot contains spaces e.g. -dr MyAppDirRef).
-ext <extension> | Extension assembly or "class, assembly".
-generate | Specify what elements to generate, one of: components, container, payloadgroup, layout (default is components).
-gg | Generate guids now. All components are given a guid when heat is run.
-g1 | Generate component guids without curly braces.
-ke | Keep empty directories.
-nologo | Skip printing heat logo information.
-out | Specify output file (default: write to current directory).
-platform | Platform to set when harvesting the project.
-pog:<group> | Specify output group of Visual Studio project, one of: Binaries, Symbols, Documents, Satellites, Sources, Content.
-projectname | Overridden project name to use in variables.
-scom | Suppress COM elements.
-sfrag | Suppress generation of fragments for directories and components.
-srd | Suppress harvesting the root directory as an element.
-sreg | Suppress registry harvesting.
-suid | Suppress unique identifiers for files, components, & directories.
-svb6 | Suppress VB6 COM registration entries. 
-sw<N> | Suppress all warnings or a specific message ID, e.g. -sw1011 -sw1012.
-t <xsl> | Transform harvested output with XSL file.
-indent <n> | Indentation multiple (overrides default of 4).
-v | Verbose output.
-var <VariableName> | Substitute File/@Source="SourceDir" with a preprocessor or a wix variable 
-wixvar | Generate binder variables instead of preprocessor variables.
-wx[N] | Treat all warnings or a specific message ID as an error. e.g. -wx1011 -wx1012.
-template <template> | Use template, one of fragment, module, product.

### CODE REVIEW

**Comment by devlead** [comment]
@RLittlesII FYI Tests are failing on Linux / OSX
https://travis-ci.org/cake-build/cake/jobs/130487778#L1359

**Comment by RLittlesII** [comment]
@devlead still WIP.  All tests passed locally with ``` .\build.ps1 -Target Run-Unit-Tests ``` last night.  Now I am getting an issue trying to load System.Linq 4.0.0.0 and my nuget version being under 3.0.  I'll let you know when I am ready for review.  Thanks for the heads up!

**Comment by devlead** [comment]
@RLittlesII superb :+1:

**Comment by RLittlesII** [comment]
@devlead having local issues with the build script.  Unable to execute ```build.ps1``` on my Windows 10 box.  Get an issue installing Polly.  Tells me my nuget version is too low, its 3.4.3 thinks its 2.8.6.  Pushed so I can pull down to my Win 10 VM to finish up the PR.  :grimacing:

**Comment by gep13** [comment]
@RLittlesII try clearing out the Tools folder completely, and run it again.

**Comment by RLittlesII** [comment]
@gep13 I get the same issue on my Mac. Ran `nuget update -self`  I'll give it a go tomorrow morning, thanks!

**Comment by RLittlesII** [comment]
I believe this is ready for review.  Removing WIP.

File: src/Cake.Common.Tests/Fixtures/Tools/WiX/HeatFixture.cs [hunk 1]
```diff
@@ -0,0 +1,25 @@
+using System.Collections.Generic;
```
Comment (gep13) [inline src/Cake.Common.Tests/Fixtures/Tools/WiX/HeatFixture.cs]: @RLittlesII can you add the new license file header?  You will find an example on all the other files in the repo.  This is a new requirement after joining the .Net Foundation.

File: src/Cake.Common.Tests/Unit/Tools/WiX/HeatRunnerTests.cs [hunk 2]
```diff
@@ -0,0 +1,674 @@
+using System;
```
Comment (gep13) [inline src/Cake.Common.Tests/Unit/Tools/WiX/HeatRunnerTests.cs]: @RLittlesII can you add the new license file header? You will find an example on all the other files in the repo. This is a new requirement after joining the .Net Foundation.

File: src/Cake.Common.Tests/Unit/Tools/WiX/HeatRunnerTests.cs [hunk 3]
```diff
@@ -0,0 +1,674 @@
...
+using System.Collections.Generic;
+using Cake.Common.Tests.Fixtures.Tools.WiX;
+using Cake.Common.Tools.WiX.Heat;
+using Cake.Core;
+using Cake.Core.IO;
+using Cake.Testing;
+using Xunit;
```
Comment (gep13) [inline src/Cake.Common.Tests/Unit/Tools/WiX/HeatRunnerTests.cs]: @RLittlesII this should be the last using statement. Sorted alphabetically.

File: src/Cake.Common.Tests/Fixtures/Tools/WiX/HeatFixture.cs (hunk 1 above)
Comment (RLittlesII) [inline src/Cake.Common.Tests/Fixtures/Tools/WiX/HeatFixture.cs]: Will do.

File: src/Cake.Common.Tests/Unit/Tools/WiX/HeatRunnerTests.cs (hunk 2 above)
Comment (RLittlesII) [inline src/Cake.Common.Tests/Unit/Tools/WiX/HeatRunnerTests.cs]: Will do.

File: src/Cake.Common/Tools/WiX/Heat/GenerateType.cs [hunk 4]
```diff
@@ -0,0 +1,28 @@
+﻿namespace Cake.Common.Tools.WiX.Heat
```
Comment (gep13) [inline src/Cake.Common/Tools/WiX/Heat/GenerateType.cs]: @RLittlesII can you add the new license file header? You will find an example on all the other files in the repo. This is a new requirement after joining the .Net Foundation.

File: src/Cake.Common/Tools/WiX/Heat/GenerateType.cs [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+{
+    /// <summary>
+    /// Type of elements to generate
+    /// </summary>
+    public enum GenerateType
```
Comment (gep13) [inline src/Cake.Common/Tools/WiX/Heat/GenerateType.cs]: @RLittlesII Can we rename this to WiXGenerateType?

File: src/Cake.Common/Tools/WiX/Heat/HarvestType.cs [hunk 6]
```diff
@@ -0,0 +1,38 @@
+﻿namespace Cake.Common.Tools.WiX.Heat
```
Comment (gep13) [inline src/Cake.Common/Tools/WiX/Heat/HarvestType.cs]: @RLittlesII can you add the new license file header? You will find an example on all the other files in the repo. This is a new requirement after joining the .Net Foundation.

File: src/Cake.Common/Tools/WiX/Heat/HarvestType.cs [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+{
+    /// <summary>
+    /// The type of object file to harvest from.
+    /// </summary>
+    public enum HarvestType
```
Comment (gep13) [inline src/Cake.Common/Tools/WiX/Heat/HarvestType.cs]: @RLittlesII can we rename this to WiXHarvestType

File: src/Cake.Common/Tools/WiX/Heat/HeatRunner.cs [hunk 8]
```diff
@@ -0,0 +1,307 @@
...
+                builder.Append(settings.Indent.ToString());
+            }
+
+            // Verbose
+            if (settings.Verbose)
+            {
+                builder.Append("-v");
+            }
+
+            // Preprosessor variable
```
Comment (gep13) [inline src/Cake.Common/Tools/WiX/Heat/HeatRunner.cs]: @RLittlesII slight typo here

File: src/Cake.Common/Tools/WiX/Heat/HeatSettings.cs [hunk 9]
```diff
@@ -0,0 +1,252 @@
...
+        /// </summary>
+        public IEnumerable<string> Extensions { get; set; }
+
+        /// <summary>
+        /// Gets or sets a value indicating whether [no logo].
+        /// </summary>
+        /// <value>
+        ///   <c>true</c> if [no logo]; otherwise, <c>false</c>.
+        /// </value>
+        public bool NoLogo { get; set; }
```
Comment (gep13) [inline src/Cake.Common/Tools/WiX/Heat/HeatSettings.cs]: @RLittlesII some whitespace in this file:

image

File: src/Cake.Common/Tools/WiX/Heat/HeatSettings.cs [hunk 10]
```diff
@@ -0,0 +1,252 @@
+using System.Collections.Generic;
```
Comment (gep13) [inline src/Cake.Common/Tools/WiX/Heat/HeatSettings.cs]: @RLittlesII can you add the new license file header? You will find an example on all the other files in the repo. This is a new requirement after joining the .Net Foundation.

File: src/Cake.Common/Tools/WiX/Heat/HeatSettings.cs [hunk 11]
```diff
@@ -0,0 +1,252 @@
...
+        /// </value>
+        public string Transform { get; set; }
+
+        /// <summary>
+        /// Gets or sets the file.
+        /// </summary>
+        /// <value>
+        /// The file.
+        /// </value>
+        public FilePath File { get; set; }
```
Comment (gep13) [inline src/Cake.Common/Tools/WiX/Heat/HeatSettings.cs]: @RLittlesII this settings never seems to be getting used.  Is it required?

File: src/Cake.Common/Tools/WiX/Heat/HeatSettings.cs [hunk 12]
```diff
@@ -0,0 +1,252 @@
...
+        /// </value>
+        public bool SuppressCom { get; set; }
+
+        /// <summary>
+        /// Gets or sets a value indicating whether [suppress registry].
+        /// </summary>
+        /// <value>
+        ///   <c>true</c> if [suppress registry]; otherwise, <c>false</c>.
+        /// </value>
+        public bool SuppressRegistry { get; set; }
```
Comment (gep13) [inline src/Cake.Common/Tools/WiX/Heat/HeatSettings.cs]: @RLittlesII this setting never seems to be getting used.  Is it required?

File: src/Cake.Common/Tools/WiX/Heat/HeatSettings.cs [hunk 13]
```diff
@@ -0,0 +1,252 @@
...
+        /// <summary>
+        /// Gets or sets the configuration to set when harvesting the project.
+        /// </summary>
+        /// <value>
+        /// The configuration.
+        /// </value>
+        public string Configuration { get; set; }
+
+        /// <summary>
+        /// Gets or sets the overriden directory identifier for generated directory elements.
```
Comment (gep13) [inline src/Cake.Common/Tools/WiX/Heat/HeatSettings.cs]: @RLittlesII slight typo here, should be overridden.

File: src/Cake.Common/Tools/WiX/Heat/OutputGroupType.cs [hunk 14]
```diff
@@ -0,0 +1,38 @@
+﻿namespace Cake.Common.Tools.WiX.Heat
```
Comment (gep13) [inline src/Cake.Common/Tools/WiX/Heat/OutputGroupType.cs]: @RLittlesII can you add the new license file header? You will find an example on all the other files in the repo. This is a new requirement after joining the .Net Foundation.

File: src/Cake.Common/Tools/WiX/Heat/TemplateType.cs [hunk 15]
```diff
@@ -0,0 +1,23 @@
+﻿namespace Cake.Common.Tools.WiX.Heat
```
Comment (gep13) [inline src/Cake.Common/Tools/WiX/Heat/TemplateType.cs]: @RLittlesII can you add the new license file header? You will find an example on all the other files in the repo. This is a new requirement after joining the .Net Foundation.

File: src/Cake.Common/Tools/WiX/WiXAliases.cs [hunk 16]
```diff
@@ -4,6 +4,7 @@
 using System;
```
Comment (gep13) [inline src/Cake.Common/Tools/WiX/WiXAliases.cs]: @RLittlesII can you add the new license file header? You will find an example on all the other files in the repo. This is a new requirement after joining the .Net Foundation.

Comment (gep13) [inline src/Cake.Common/Tools/WiX/WiXAliases.cs]: @RLittlesII can you add the new license file header? You will find an example on all the other files in the repo. This is a new requirement after joining the .Net Foundation.

File: src/Cake.Common/Tools/WiX/Heat/OutputGroupType.cs [hunk 17]
```diff
@@ -0,0 +1,38 @@
...
+{
+    /// <summary>
+    /// The Output Group of Visual Studio project
+    /// </summary>
+    public enum OutputGroupType
```
Comment (gep13) [inline src/Cake.Common/Tools/WiX/Heat/OutputGroupType.cs]: @RLittlesII can this be WiXOutputGroupType?

File: src/Cake.Common/Tools/WiX/Heat/TemplateType.cs [hunk 18]
```diff
@@ -0,0 +1,23 @@
...
+{
+    /// <summary>
+    /// Template type to use for harvesting.
+    /// </summary>
+    public enum TemplateType
```
Comment (gep13) [inline src/Cake.Common/Tools/WiX/Heat/TemplateType.cs]: @RLittlesII can this be WiXTemplateType?

File: src/Cake.Common/Tools/WiX/WiXAliases.cs [hunk 19]
```diff
@@ -147,5 +148,107 @@ public static void WiXLight(this ICakeContext context, IEnumerable<FilePath> obj
             var runner = new LightRunner(context.FileSystem, context.Environment, context.ProcessRunner, context.Tools);
             runner.Run(objectFiles, settings ?? new LightSettings());
         }
+
+        /// <summary>
+        /// Harvests all files that match the pattern.
+        /// </summary>
+        /// <example>
+        /// <code>
```
Comment (gep13) [inline src/Cake.Common/Tools/WiX/WiXAliases.cs]: @RLittlesII can we get some example usage in here?

File: src/Cake.Common/Tools/WiX/WiXAliases.cs [hunk 20]
```diff
@@ -147,5 +148,107 @@ public static void WiXLight(this ICakeContext context, IEnumerable<FilePath> obj
...
+            }
+
+            WiXHeat(context, files, new HeatSettings());
+        }
+
+        /// <summary>
+        /// Harvests all files that match the pattern.
+        /// </summary>
+        /// <example>
+        /// <code>
```
Comment (gep13) [inline src/Cake.Common/Tools/WiX/WiXAliases.cs]: @RLittlesII can we get some example usage here?

File: src/Cake.Common/Tools/WiX/WiXAliases.cs [hunk 21]
```diff
@@ -147,5 +148,107 @@ public static void WiXLight(this ICakeContext context, IEnumerable<FilePath> obj
...
+            }
+
+            WiXHeat(context, files, settings ?? new HeatSettings());
+        }
+
+        /// <summary>
+        /// Harvests files in the provided object files.
+        /// </summary>
+        /// <example>
+        /// <code>
```
Comment (gep13) [inline src/Cake.Common/Tools/WiX/WiXAliases.cs]: @RLittlesII can we get some example usage here?

File: src/Cake.Common/Tools/WiX/WiXAliases.cs [hunk 22]
```diff
@@ -147,5 +148,107 @@ public static void WiXLight(this ICakeContext context, IEnumerable<FilePath> obj
...
+
+            var runner = new HeatRunner(context.FileSystem, context.Environment, context.ProcessRunner, context.Tools);
+            runner.Run(objectFiles, new HeatSettings());
+        }
+
+        /// <summary>
+        /// Harvests files in the provided object files.
+        /// </summary>
+        /// <example>
+        /// <code>
```
Comment (gep13) [inline src/Cake.Common/Tools/WiX/WiXAliases.cs]: @RLittlesII can we get some example usage here?

**Comment by gep13** [comment]
@RLittlesII sorry it has taken so long for us to get round to looking at this!  Thank you for your contribution!  I have just taken a look at this, and I have left some comments inline.  If you can take a look through them, and let me know what you think, and fix up what you agree with, I can then take another look.  Thanks again!

**Comment by RLittlesII** [comment]
@gep13 I'll take a look at it this evening.  No worries.  I know the cake team is busy. :+1:

**Comment by RLittlesII** [comment]
@gep13 Realized an implementation issue.  I think I need to be taking in the directory that WiX Heat points to and the output file name as parameters in the alias.  heat.exe fails immediately if the -out isn't present.  Going to clean this up and push more changes.

**Comment by gep13** [comment]
@RLittlesII I have taken the liberty of merging this into develop here: https://github.com/cake-build/cake/commit/44fa0d8a5aeb6356a3eb95d5c830c0fc44678276 rather than making you do another rebase.  Thanks again for this contribution.

**Pull request merged**